### PR TITLE
Tile Simplification

### PR DESCRIPTION
**Pull request by cquiroz**
Tiles can render controls on the title that may interact with the body.
At the moment this can be done in two ways which is confusing and it is a bit hard to share state
This PR changes this, now a Tile can take two components for body and tile (providing a single way to do this) and a mechanism for them to share state.

This is a relatively large PR and though it seems to add lines it is clearer to understand now as components for tile and body are clearly separated instead of being intermeshed. Also we were able to move some the shared state down.

I tried to keep a naming convention such that we can more or less easily identify the role of each component (Body/Title)

### CODE REVIEW

File: common/src/main/scala/explore/components/Tile.scala [hunk 1]
```diff
@@ -15,57 +17,103 @@ import lucuma.react.common.ReactFnProps
...
   controllerClass:   Css = Css.Empty, // applied to wrapping div when in a TileController.
   bodyClass:         Css = Css.Empty, // applied to tile body
   tileClass:         Css = Css.Empty, // applied to the tile
   tileTitleClass:    Css = Css.Empty  // applied to the title
-)(val render: Tile.RenderInTitle => VdomNode)
-    extends ReactFnProps[Tile](Tile.component) {
-  def showMaximize: Boolean =
-    state === TileSizeState.Minimized || (canMaximize && state === TileSizeState.Minimized)
+)(
+  val tileBody:      View[A] => VdomNode,
```
Comment (cquiroz) [inline common/src/main/scala/explore/components/Tile.scala:76]: This is the key change, components for body and title are passed as a param

File: common/src/main/scala/explore/components/Tile.scala [hunk 2]
```diff
@@ -75,47 +123,44 @@ object Tile:
...
                     minimizeButton.when(p.showMinimize),
                     maximizeButton.when(p.showMaximize)
-              )
+              ).unless(p.fullSize)
             ),
             // Tile body
-            infoRef.value
-              .map(node =>
-                <.div(ExploreStyles.TileBody |+| p.bodyClass,
-                      p.render(info => ReactPortal(info, node))
```
Comment (cquiroz) [inline common/src/main/scala/explore/components/Tile.scala:116]: No need for a Portal anymore

File: explore/src/main/scala/explore/tabs/ObsTabTiles.scala [hunk 3]
```diff
@@ -263,20 +261,11 @@ object ObsTabTiles:
             .whenA(itcProps.isExecutable)
             .runAsyncAndForget
       }
-      // ITC selected target. Here to be shared by the ITC tile body and title
```
Comment (cquiroz) [inline explore/src/main/scala/explore/tabs/ObsTabTiles.scala:266]: This is an example of a state that was pushed down to the tile where it makes more sense

File: explore/src/main/scala/explore/attachments/ObsAttachmentsTableBody.scala [hunk 4]
```diff
@@ -323,54 +327,17 @@ object ObsAttachmentsTable extends ObsAttachmentUtils:
...
           getRowId = (row, _, _) => RowId(row.get.id.toString),
           meta = TableMeta(client, props.obsAttachmentAssignments, urlMap.get, props.readOnly)
         )
-      .useStateView(Enumerated[ObsAttachmentType].all.head)
-      .render: (props, ctx, client, action, _, _, _, table, newAttType) =>
-        import ctx.given
-
+      .render: (props, _, _, _, _, _, table) =>
         React.Fragment(
-          props.renderInTitle(
```
Comment (cquiroz) [inline explore/src/main/scala/explore/attachments/ObsAttachmentsTableBody.scala:340]: This is an example where a new component for the title was extracted out from the body and it is now an independent component for the title

File: explore/src/main/scala/explore/components/ColumnSelectorInTitle.scala [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+import lucuma.react.common.ReactFnProps
+import lucuma.react.table.ColumnId
+import lucuma.react.table.Table
+import lucuma.ui.table.ColumnSelector
+
+case class ColumnSelectorState[A, B](
+  table: Option[Table[A, B]] = None
+)
+
+case class ColumnSelectorInTitle[A, B](
```
Comment (cquiroz) [inline explore/src/main/scala/explore/components/ColumnSelectorInTitle.scala:20]: This was extracted to be common across several tiles

File: common/src/main/scala/explore/components/Tile.scala [hunk 6]
```diff
@@ -15,57 +17,103 @@ import lucuma.react.common.ReactFnProps
 import lucuma.react.common.style.*
 import lucuma.react.primereact.Button
 import lucuma.ui.syntax.all.given
-import org.scalajs.dom
 
-case class Tile(
+/**
```
Comment (rpiaggio) [inline common/src/main/scala/explore/components/Tile.scala:21]: Thanks for the doc.

File: explore/src/main/scala/explore/components/ColumnSelectorInTitle.scala (hunk 5 above)
Comment (rpiaggio) [reply explore/src/main/scala/explore/components/ColumnSelectorInTitle.scala:20]: Nice

File: explore/src/main/scala/explore/attachments/ObsAttachmentsTableBody.scala [hunk 7]
```diff
@@ -60,17 +60,22 @@ import org.typelevel.log4cats.Logger
 
 import scala.collection.immutable.SortedSet
 
-case class ObsAttachmentsTable(
+object ObsAttachmentsTableTileState extends NewType[Action]:
+  def apply(): ObsAttachmentsTableTileState = ObsAttachmentsTableTileState(Action.None)
+type ObsAttachmentsTableTileState = ObsAttachmentsTableTileState.Type
+
+case class ObsAttachmentsTableBody(
```
Comment (rpiaggio) [inline explore/src/main/scala/explore/attachments/ObsAttachmentsTableBody.scala:67]: Maybe we want to keep the name of the file `ObsAttachmentsTable` (without `Body`).

Also, have you considered the following structure?
``` scala
object ObsAttachmentsTable:
  object State extends NewType[Action]
  
  case class Body(...)
  object Body {
    ...
  }
  
  case class Title(...)
  object Title {
    ...
  }
```

**Review by rpiaggio — APPROVED** [review]
I haven't gone in detail into all the individual tiles, but the general concept LGTM.

Thank you for tackling this!

File: explore/src/main/scala/explore/attachments/ObsAttachmentsTableBody.scala (hunk 7 above)
Comment (cquiroz) [reply explore/src/main/scala/explore/attachments/ObsAttachmentsTableBody.scala:67]: Seems good, maybe calling it `ObsAttachmentsTile`?

**Comment by cquiroz** [comment]
Thanks to @rpiaggio for the great ideas on how to implement this

**Review by toddburnside — APPROVED** [review]
I must admit my eyes glazed over in some of the tile refactoring... But this looks great. It also look like a lot of work. Thanks for doing this.

**Pull request merged**